engine/resources/hetzner.go

```diff
@@ -0,0 +1,790 @@
+		Image:            image,             // *Image
+		SSHKeys:          keylist,           // []*SSHKey
+		Location:         nil,               // *Location
+		Datacenter:       datacenter,        // *Datacenter
+		UserData:         "",                // string
+		StartAfterCreate: &startAfterCreate, // *bool
+		Labels:           nil,               // map[string]string
+		Automount:        &automount,        // *bool
+		Volumes:          nil,               // []*Volume
+		Networks:         nil,               // []*Network
+		Firewalls:        nil,               // []*ServerCreateFirewall
+		PlacementGroup:   nil,               // *PlacementGroup
+	}
+	// validate resulting config with hloud package
+	if err := hcloud.ServerCreateOpts.Validate(obj.serverconfig); err != nil {
+		return errwrap.Wrapf(err, "invalid serverconfig")
+	}
+	return nil
+}
+
+// getRescueKeys builds a list of keys to be enabled for rescue mode over ssh.
+// HetznerRes.ServerRescueSSHKeys provides the selected keys as []string by
+// name. The corresponding data is collected with the hetzner client (if valid).
+// The resulting keylist is formatted as []*hcloud.SSHKey for use with hcloud,
+// and saved for later use in private field HetznerRes.rescuekeys.
+func (obj *HetznerRes) getRescueKeys(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getRescueKeys()")
+	}
+	var keylist []*hcloud.SSHKey
+	for _, keyname := range obj.ServerRescueSSHKeys {
+		key, _, err := obj.client.SSHKey.GetByName(ctx, keyname)
+		if err != nil {
+			return errwrap.Wrapf(err, "SSHKey GetByName(%s) failed", keyname)
+		}
+		if key == nil {
+			return fmt.Errorf("unknown keyname: %s", keyname)
+		}
+		if obj.init.Debug {
+			obj.init.Logf("append known keyname: %s", key)
+		}
+		keylist = append(keylist, key)
+	}
+	obj.rescuekeys = keylist
+	return nil
+}
+
+// waitFor provides a general function that returns nil when the exit condition
+// corresponding to the input string is satisfied. It retries every WaitInterval
+// until the condition is finally satisfied, the WaitTimeout is reached, the
+// context is cancelled, or a check fails and returns an error. The supported
+// exit conditions are referred to as `convergence`, `steadystate`, `creation`,
+// `deletion` and `rescuemode`. Each has its own logical condition related to
+// the resource params and whether or not the API confirms intentions.
+func (obj *HetznerRes) waitFor(ctx context.Context, exitCondition string) error {
+	if obj.init.Debug {
+		obj.init.Logf("waitFor(exitCondition: %s)", exitCondition)
+	}
+	var err error = nil
+	conditionSatisfied := false
+	timeout := time.After(time.Duration(obj.WaitTimeout) * time.Second)
+	for {
+		// Get up-to-date serverinfo.
+		if err := obj.getServerUpdate(ctx); err != nil {
+			return errwrap.Wrapf(err, "failed serverupdate request")
+		}
+		// Check if the appropriate exit condition is satisfied.
+		switch exitCondition {
+		case "convergence":
```

Comment:
Can you use constants instead of these magic strings?

Reply:
Yes of course. I was a bit hesitant to do it this way because the constants seemed out of place as global values for the resource. Alternatively, I could pass the exitCondition as functions, maybe that's even cleaner? Then it's immediately clear which exit condition is being used by the waiter, without having that intermediate string.

Reply:
I tried the functional approach, and personally I really like it. Let me know what you think. A wait statement now looks something like `err := obj.WaitFor(ctx, serverInSteadyState)` where `serverInSteadyState` is a function that checks whether the server is in steady state. I added a `HetznerServerCondition` type to make sure any functions of that conditional structure can be used as needed.